tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -7630,16 +7630,78 @@ void SortOp::getCanonicalizationPatterns(RewritePatternSet& results,
 // TransposeOp
 //===----------------------------------------------------------------------===//
 
+template <typename T>
+static DenseElementsAttr foldTransposeHelper(mhlo::TransposeOp op,
+                                             DenseElementsAttr valueAttr) {
+  llvm::SmallVector<int64_t> permutation =
+      llvm::to_vector(op.getPermutation().getValues<int64_t>());
+  int64_t rank = permutation.size();
+  auto inputShape = op.getOperand().getType().cast<ShapedType>().getShape();
+  auto outputType = op.getType().cast<ShapedType>();
+  auto outputShape = outputType.getShape();
+
+  assert(outputType.hasStaticShape() && "output shape must be static");
+  assert(rank >= 2 && "permutation size must >= 2");
+  llvm::SmallVector<int64_t> strides(rank, 1);
+  llvm::SmallVector<int64_t> outputStrides(rank, 1);
+  llvm::SmallVector<int64_t> outputPermutation(rank, -1);
+  for (int64_t i = rank - 2; i >= 0; i--) {
+    strides[i] = strides[i + 1] * inputShape[i + 1];
+    outputStrides[i] = outputStrides[i + 1] * outputShape[i + 1];
+  }
+  for (int64_t i = 0; i < rank; i++) {
+    outputPermutation[permutation[i]] = i;
+  }
+
+  auto calculateOutputIndices = [&](int64_t index) -> SmallVector<int64_t> {
+    SmallVector<int64_t> indices(rank, -1);
+    for (int64_t i = 0; i < rank; i++) {
+      indices[i] = index / outputStrides[i];
+      index = index % outputStrides[i];
+    }
+    return indices;
+  };
+
+  SmallVector<T> values;
+  for (int64_t i = 0; i < outputType.getNumElements(); i++) {
+    auto indices = calculateOutputIndices(i);
+    int64_t inputIndex = 0;
+    for (int64_t k = 0; k < rank; k++) {
+      inputIndex += indices[k] * strides[outputPermutation[k]];
+    }
+    values.push_back(*(valueAttr.getValues<T>().begin() + inputIndex));
```

Comment:
It might simplify things to prepare the entire vector first and then run something like `for (auto idxAndValue : llvm::enumerate(valueAttr.getValues()))`.

Otherwise, please reserve space at the beginning for `values` as we know how big it will be and don't need to pay the reallocation cost.